File: typescript/src/item.list.client.ts

import { ItemList, Item, Priced } from './item';
import { sendRequest } from './api.client';
import * as env from './env';

export class ItemListClient implements ItemList {

    async get(code: string): Promise<Item & Priced | undefined> {
        try {
            const result = await sendRequest(`http://${env.ITEM_API_BASE_URL}:${env.ITEM_API_PORT}/item/${code}`);
            return <Item & Priced>result;
        } catch (error) {
            if (error === 404) {
                return undefined;
            }
            throw error;
        }
    }

    async add(_: Item): Promise<void> {
        throw new Error('Method not implemented.');
    }

    async includes(_: Item): Promise<boolean> {
        throw new Error('Method not implemented.');
    }

}

File: haskell/ItemListClient.hs

{-# LANGUAGE OverloadedStrings #-}

module ItemListClient ( ItemList, createItemList, getItem ) where

  import Data.Aeson               (decode)
  import Network.HTTP.Client      (newManager, defaultManagerSettings, parseRequest, httpLbs, method, responseBody)
  import Item ( Item )

  data ItemList = ItemList { baseUrl :: String }

  class ItemListClass itemList where
    getItem :: itemList -> String -> IO [Item]

  createItemList :: String -> ItemList
  createItemList baseURL = ItemList { baseUrl = baseURL }

  instance ItemListClass ItemList where
    getItem itemList code = do
      manager                       <- newManager defaultManagerSettings
      initialRequest                <- parseRequest $ baseUrl itemList ++ "/item/" ++ code
      let request                   = initialRequest { method = "GET" }
      response                      <- httpLbs request manager
      let item                      = decode $ responseBody response :: Maybe Item
      case item of
        Just foundItem                -> return [foundItem]
        _                             -> return []
